pkg/ibmvpc/subnetsIPBlocks.go

```diff
@@ -0,0 +1,237 @@
+/*
+Copyright 2023- IBM Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+package ibmvpc
+
+import (
+	"fmt"
+
+	vpc1 "github.com/IBM/vpc-go-sdk/vpcv1"
+
+	"github.com/np-guard/cloud-resource-collector/pkg/ibm/datamodel"
+	"github.com/np-guard/models/pkg/ipblock"
+)
+
+// ///////////////////////////////////////////////////////////////////////
+// connection from and to load balancer are done via private ips
+// created to it upon its deployment.
+// However, when a load balancer is deployed,
+// Private IPs are not created for all the load balancer subnets.
+// To understand and monitor the connectivity that is potentially induced by the load balancer,
+// we create a private IP for all of the load balancer subnets.
+// (we calls these private IPs fake private Ips)
+// Moreover, if a filter rule splits the subnet's cidr to few blocks, we want to create private ip per each such block.
+
+// To create a private IP which does not exist in the config, we need an unused address.
+// subnetsIPBlocks has two main purposes:
+//   1. for each subnet, it calculates splitByFiltersBlocks - the atomic blocks induced by the filters.
+//      splitByFiltersBlocks are such that:
+//   		a. the blocks are disjoint
+//	    	b. the union of thr blocks is the subnet cidr
+//      we calculate splitByFiltersBlocks of a subnet by:
+//          (a) collecting all the filters' rules blocks.
+//          (b) adding allCidr block to these blocks.
+//          (c) from these blocks, create a list of disjoint blocks
+//          (d) for each of these blocks, intersect the block with the subnet cidr
+//          (e) collect all the non empty intersections we get in (d)
+//   2. allocate a free address for the fake private IPs - to this end, we holds for each subnet freeAddressesBlocks,
+//      freeAddressesBlocks are splitByFiltersBlocks minus all the addresses that was already allocated.
+//      to get freeAddressesBlocks, we first copy splitByFiltersBlocks, than we remove the subnet already allocated addresses.
+//      (we get these addresses from the subnet reserved IP).
+//       when a free address is needed, we take the first address of the block and remove it from the free blocks list.
+
+// /////////////////////////////////////////////////////////////////////////////////////////
+// subnetIPBlocks holds the blocks of a subnet
+type subnetIPBlocks struct {
+	subnetOriginalBlock  *ipblock.IPBlock // the block of the original cidr of the subnet
+	splitByFiltersBlocks []*ipblock.IPBlock
+	freeAddressesBlocks  []*ipblock.IPBlock // each block in splitByFiltersBlocks has a corresponding block at freeAddressesBlocks
+}
+type subnetsIPBlocks map[string]*subnetIPBlocks
```

Comment:
add documentation to the type `subnetsIPBlocks`, such as `subnetsIPBlocks is a map from subnet UID to ... `

Reply:
also consider renaming ,as `subnetIPBlocks` and `subnetsIPBlocks` are too similar and confusing

Reply:
done